File: leetcode/python/907_sum_of_subarray_minimums.py
class Solution:
    def sumSubarrayMins(self, arr: list[int]) -> int:
        MOD = 10 ** 9 + 7
        stack = []
        res = 0;

        for i in range(len(arr) + 1):
            
            # when i reaches the array length, it is an indication that
            # all the elements have been processed, and the remaining
            # elements in the stack should now be popped out.

            while stack and (i == len(arr) or arr[stack[-1]] >= arr[i]):
                print(stack)

                # Notice the sign ">=", This ensures that no contribution
                # is counted twice. right_boundary takes equal or smaller 
                # elements into account while left_boundary takes only the
                # strictly smaller elements into account

                mid = stack.pop()
                left_boundary = -1 if not stack else stack[-1]
                right_boundary = i

                # count of subarrays where mid is the minimum element
                count = (mid - left_boundary) * (right_boundary - mid)
                res += (count * arr[mid])

            stack.append(i)
        
        return res % MOD

    
s = Solution()
print(s.sumSubarrayMins([3,1,2,4]))
# print(s.sumSubarrayMins([11,81,94,43,3]))

File: leetcode/rust/src/907_sum_of_subarray_minimums.rs
impl Solution {
    // [1,2,3,4]
    // 
    pub fn sum_subarray_mins(arr: Vec<i32>) -> i32 {
        let mut res = 0;
        let mut stack: Vec<usize> = vec![];
        let MOD = 1_000_000_007;

        for r in 0..=arr.len() {
            while !stack.is_empty() && (r == arr.len() || arr[*stack.last().unwrap()] >= arr[r]) {
                let m: i32 = stack.pop().unwrap() as i32;
                let l: i32 = if stack.is_empty() {
                    -1
                } else {
                    *stack.last().unwrap() as i32
                };
                let count: i32 = (r as i32 - m) * (m - l);
                res = (count as usize * arr[m as usize] as usize + res) % MOD;
            }
            stack.push(r);
        }

        res as i32
    }
}
